build failures
```

======================================================================
FAIL: Ensure dates are transformed for excel when `transform_dates` is set to True
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/export/tests/test_get_export_file.py", line 302, in test_transform_dates
    u'rows': [[None], [couch_to_excel_datetime('2015-07-22T14:16:49.584880Z', None)]],
AssertionError: {u'My table': {u'headers': [u'Date'], u'rows': [[u'---'], [u'2015-07-22 14:16:49 [truncated]... != {u'My table': {u'headers': [u'Date'], u'rows': [[None], ['2015-07-22 14:16:49']] [truncated]...
  {u'My table': {u'headers': [u'Date'],
-                u'rows': [[u'---'], [u'2015-07-22 14:16:49']]}}
?                           ^^^^^^    -
+                u'rows': [[None], ['2015-07-22 14:16:49']]}}
?                           ^^^^

======================================================================
FAIL: test_export_transforms (corehq.apps.export.tests.test_get_export_file.ExportTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock.py", line 1224, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/export/tests/test_get_export_file.py", line 589, in test_export_transforms
    [u'2016-04-27'],  # offset by 3 since that's the mocked random offset
AssertionError: {u'My table': {u'headers': [u'DEID Date Transform column [sensitive]'], u'rows': [truncated]... != {u'My table': {u'headers': [u'DEID Date Transform column [sensitive]'], u'rows': [truncated]...
  {u'My table': {u'headers': [u'DEID Date Transform column [sensitive]'],
-                u'rows': [[u'---'], [u'2016-04-07'], [u'2016-04-27']]}}
?                           ^^^^^^
+                u'rows': [[None], [u'2016-04-07'], [u'2016-04-27']]}}
?                           ^^^^
 



```

--- a/corehq/apps/export/const.py
+++ b/corehq/apps/export/const.py
@@ -52,3 +52,8 @@ PROPERTY_TAG_STOCK = 'stock'
 FORM_EXPORT = 'form'
 CASE_EXPORT = 'case'
 MAX_EXPORTABLE_ROWS = 100000
+
+# When a question is missing completely from a form/case this should be the value
+MISSING_VALUE = '---'
+# When a question has been answered, but is blank, this shoudl be the value
+EMPTY_VALUE = ''

--- a/corehq/apps/export/models/new.py
+++ b/corehq/apps/export/models/new.py
@@ -51,6 +51,8 @@ from corehq.apps.export.const import (
     USER_DEFINED_SPLIT_TYPES,
     PLAIN_USER_DEFINED_SPLIT_TYPE,
     DATA_SCHEMA_VERSION,
+    MISSING_VALUE,
+    EMPTY_VALUE,
 )
 from corehq.apps.export.exceptions import BadExportConfiguration
 from corehq.apps.export.dbaccessors import (
@@ -206,6 +208,8 @@ class ExportColumn(DocumentSchema):
             except ValueError:
                 # Unable to convert the string to a date
                 pass
+        if value is None:
+            value = MISSING_VALUE
         return value
 
     @staticmethod
@@ -1530,7 +1534,7 @@ class MultiMediaExportColumn(ExportColumn):
     def get_value(self, domain, doc_id, doc, base_path, transform_dates=False, **kwargs):
         value = super(MultiMediaExportColumn, self).get_value(domain, doc_id, doc, base_path, **kwargs)
 
-        if not value:
+        if not value or value == MISSING_VALUE:
             return value
 
         download_url = u'{url}?attachment={attachment}'.format(
@@ -1569,7 +1573,10 @@ class SplitGPSExportColumn(ExportColumn):
         if not split_column:
             return value
 
-        values = [None] * 4
+        if value == MISSING_VALUE:
+            return [MISSING_VALUE] * 4
+
+        values = [EMPTY_VALUE] * 4
 
         if not isinstance(value, basestring):
             return values
@@ -1615,14 +1622,20 @@ class SplitExportColumn(ExportColumn):
         if not split_column:
             return value
 
+        if value == MISSING_VALUE:
+            value = [MISSING_VALUE] * len(self.item.options)
+            if not self.ignore_unspecified_options:
+                value.append(MISSING_VALUE)
+            return value
+
         if not isinstance(value, basestring):
             unspecified_options = [] if self.ignore_unspecified_options else [value]
-            return [None] * len(self.item.options) + unspecified_options
+            return [EMPTY_VALUE] * len(self.item.options) + unspecified_options
 
         selected = OrderedDict((x, 1) for x in value.split(" "))
         row = []
         for option in self.item.options:
-            row.append(selected.pop(option.value, None))
+            row.append(selected.pop(option.value, EMPTY_VALUE))
         if not self.ignore_unspecified_options:
             row.append(" ".join(selected.keys()))
         return row
@@ -1729,7 +1742,7 @@ class StockFormExportColumn(ExportColumn):
 
         value = NestedDictGetter(path[:stock_type_path_index + 1])(doc)
         if not value:
-            return None
+            return MISSING_VALUE
 
         new_doc = None
         if isinstance(value, list):
@@ -1745,7 +1758,7 @@ class StockFormExportColumn(ExportColumn):
                 new_doc = value
 
         if not new_doc:
-            return None
+            return MISSING_VALUE
 
         return self._transform(
             NestedDictGetter(path[stock_type_path_index + 1:])(new_doc),
@@ -1791,7 +1804,7 @@ class StockExportColumn(ExportColumn):
 
         # use a list to make sure the stock states end up
         # in the same order as the headers
-        values = [None] * len(self._column_tuples)
+        values = [EMPTY_VALUE] * len(self._column_tuples)
 
         for state in states:
             column_tuple = (state.product_id, state.section_id)

--- a/corehq/apps/export/tests/test_export_column.py
+++ b/corehq/apps/export/tests/test_export_column.py
@@ -6,6 +6,8 @@ from corehq.util.view_utils import absolute_reverse
 from corehq.apps.export.const import (
     PLAIN_USER_DEFINED_SPLIT_TYPE,
     MULTISELCT_USER_DEFINED_SPLIT_TYPE,
+    MISSING_VALUE,
+    EMPTY_VALUE,
 )
 from corehq.apps.export.models import (
     ExportColumn,
@@ -53,7 +55,19 @@ class SplitColumnTest(SimpleTestCase):
             doc,
             [PathNode(name='form')],
             split_column=True),
-            [1, None, "b d"]
+            [1, EMPTY_VALUE, "b d"]
+        )
+
+        doc = {}
+        self.assertEqual(
+            column.get_value(
+                'domain',
+                'docid',
+                doc,
+                [PathNode(name='form')],
+                split_column=True
+            ),
+            [MISSING_VALUE, MISSING_VALUE, MISSING_VALUE]
         )
 
     def test_get_value_numerical(self):
@@ -71,7 +85,7 @@ class SplitColumnTest(SimpleTestCase):
             doc,
             [PathNode(name='form')],
             split_column=True),
-            [None, None, 3]
+            [EMPTY_VALUE, EMPTY_VALUE, 3]
         )
 
     def test_ignore_extas(self):
@@ -89,7 +103,18 @@ class SplitColumnTest(SimpleTestCase):
             doc,
             [PathNode(name="form")],
             split_column=True),
-            [1, None],
+            [1, EMPTY_VALUE],
+        )
+        doc = {}
+        self.assertEqual(
+            column.get_value(
+                'domain',
+                'docid',
+                doc,
+                [PathNode(name='form')],
+                split_column=True
+            ),
+            [MISSING_VALUE, MISSING_VALUE]
         )
 
     def test_basic_get_headers(self):
@@ -204,8 +229,8 @@ class TestCaseIndexExportColumn(SimpleTestCase):
         item = CaseIndexItem(path=[PathNode(name='indices'), PathNode(name='RegCase')])
         col = CaseIndexExportColumn(item=item)
 
-        self.assertEqual(col.get_value('domain', 'docid', doc, []), '')
-        self.assertEqual(col.get_value('domain', 'docid', doc2, []), '')
+        self.assertEqual(col.get_value('domain', 'docid', doc, []), EMPTY_VALUE)
+        self.assertEqual(col.get_value('domain', 'docid', doc2, []), EMPTY_VALUE)
 
 
 class TestGeopointExportColumn(SimpleTestCase):
@@ -215,13 +240,13 @@ class TestGeopointExportColumn(SimpleTestCase):
             item=GeopointItem(path=[PathNode(name='form'), PathNode(name='geo')])
         )
         result = column.get_value('domain', 'docid', {'form': {'geo': '10 20'}}, [], split_column=True)
-        self.assertEqual(result, ['10', '20', None, None])
+        self.assertEqual(result, ['10', '20', EMPTY_VALUE, EMPTY_VALUE])
 
         result = column.get_value('domain', 'docid', {'form': {'geo': '10 20'}}, [], split_column=False)
         self.assertEqual(result, '10 20')
 
         result = column.get_value('domain', 'docid', {'form': {'geo': None}}, [], split_column=True)
-        self.assertEqual(result, [None, None, None, None])
+        self.assertEqual(result, [MISSING_VALUE] * 4)
 
     def test_get_headers(self):
         column = SplitGPSExportColumn(
@@ -278,7 +303,7 @@ class TestSplitExportColumn(SimpleTestCase):
             ),
         )
         result = column.get_value('domain', 'docid', {'form': {'mc': 'foo extra'}}, [], split_column=True)
-        self.assertEqual(result, [1, None, 'extra'])
+        self.assertEqual(result, [1, EMPTY_VALUE, 'extra'])
 
         result = column.get_value('domain', 'docid', {'form': {'mc': 'foo extra'}}, [], split_column=False)
         self.assertEqual(result, 'foo extra')
@@ -317,6 +342,11 @@ class TestMultiMediaExportColumn(SimpleTestCase):
                 absolute_reverse('download_attachment', args=('my-domain', '1234'))
             )
         )
+        result = column.get_value('my-domain', '1234', {'photo': None}, [PathNode(name='form')])
+        self.assertEqual(result, MISSING_VALUE)
+
+        result = column.get_value('my-domain', '1234', {'photo': ''}, [PathNode(name='form')])
+        self.assertEqual(result, '')
 
     def test_get_value_excel_format(self):
         column = MultiMediaExportColumn(

--- a/corehq/apps/export/tests/test_get_export_file.py
+++ b/corehq/apps/export/tests/test_get_export_file.py
@@ -16,6 +16,10 @@ from corehq.apps.export.export import (
     ExportFile,
     get_export_file,
 )
+from corehq.apps.export.const import (
+    MISSING_VALUE,
+    EMPTY_VALUE,
+)
 from corehq.apps.export.models import (
     TableConfiguration,
     ExportColumn,
@@ -160,7 +164,7 @@ class WriterTest(SimpleTestCase):
                 {
                     u'My table': {
                         u'headers': [u'MC | one', u'MC | two', 'MC | extra'],
-                        u'rows': [[None, 1, 'extra'], [1, 1, '']],
+                        u'rows': [[EMPTY_VALUE, 1, 'extra'], [1, 1, '']],
 
                     }
                 }
@@ -260,8 +264,8 @@ class WriterTest(SimpleTestCase):
                         u'rows': [
                             ['question-id', '2'],
                             ['question-id', '2'],
-                            [None, None],
-                            [None, None],
+                            [MISSING_VALUE, MISSING_VALUE],
+                            [MISSING_VALUE, MISSING_VALUE],
                         ],
                     }
                 }
@@ -299,7 +303,7 @@ class WriterTest(SimpleTestCase):
                 {
                     u'My table': {
                         u'headers': [u'Date'],
-                        u'rows': [[None], [couch_to_excel_datetime('2015-07-22T14:16:49.584880Z', None)]],
+                        u'rows': [[MISSING_VALUE], [couch_to_excel_datetime('2015-07-22T14:16:49.584880Z', None)]],
 
                     }
                 }
@@ -584,7 +588,7 @@ class ExportTest(SimpleTestCase):
                             u'DEID Date Transform column [sensitive]',
                         ],
                         u'rows': [
-                            [None],
+                            [MISSING_VALUE],
                             [u'2016-04-07'],
                             [u'2016-04-27'],  # offset by 3 since that's the mocked random offset
                         ],

